src/Spotify/SpotifyAPI.cpp

```diff
@@ -118,25 +133,278 @@ TResult<Playback> SpotifyAPI::getCurrentPlayback(std::string const &accessToken,
+  headers.insert(std::pair<std::string, std::string>("Authorization",
+                                                     "Bearer " + accessToken));
+  client->SetHeaders(headers);
+
+  // build query
+  std::stringstream queryStream;
+  if (device.getID() != "") {
+    queryStream << "?device_id=" << device.getID();
+  }
+
+  // build body
+  nlohmann::json body = {{"position_ms", positionMs}};
+  if (!uris.empty()) {
+    body["uris"] = uris;
+  }
+
+  client->SetTimeout(cRequestTimeout);
+  auto response =
+      client->put("/v1/me/player/play" + queryStream.str(), body.dump());
+
+  if (response.code == cNoContent) {
+    return std::nullopt;
+  }
+
+  nlohmann::json errJson;
+  try {
+    errJson = nlohmann::json::parse(response.body);
+
+    // check for error object
+    if (errJson.find("error") != errJson.end()) {
+      SpotifyError spotifyError(errJson["error"]);
+      return errorParser(spotifyError);
+    }
+
+    // if we reach here or the exception gets thrown spotify sent an unexpected
+    // message
+  } catch (...) {
+    // parse exception
+    return Error(ErrorCode::SpotifyParseError,
+                 "[SpotifyAPI] received json couldn't be parsed");
+  }
+  // if we reach here spotify sent an unexpected message
+  return Error(ErrorCode::SpotifyAPIError,
+               "Spotify sent an unexpected message");
 }
 
 Error SpotifyAPI::errorParser(SpotifyApi::SpotifyError const &error) {
   if (error.getStatus() == cHTTPUnouthorized) {
-    if (error.getMessage() == "Invalid access token") {
-      return Error(ErrorCode::AccessDenied, error.getMessage());
-    } else if (error.getMessage() == "The access token expired") {
+    if (error.getMessage().find("Invalid access token") != std::string::npos) {
       return Error(ErrorCode::AccessDenied, error.getMessage());
+    } else if (error.getMessage().find("The access token expired") !=
+               std::string::npos) {
+      return Error(ErrorCode::SessionExpired, error.getMessage());
     }
+  } else if (error.getStatus() == cHTTPNotFound) {
+    return Error(ErrorCode::SpotifyNotFound, error.getMessage());
+  } else if (error.getStatus() == cHTTPForbidden) {
+    return Error(ErrorCode::SpotifyForbidden, error.getMessage());
   } else {
     // unhandled spotify error
-    return Error(ErrorCode::NotImplemented, error.getMessage());
+    LOG(ERROR) << "[SpotifyAPI]: Error " << error.getMessage()
+               << " Status: " << error.getStatus() << std::endl;
+    return Error(ErrorCode::SpotifyAPIError, error.getMessage());
+  }
+}
+
+std::string SpotifyAPI::stringUrlEncode(std::string const &str) {
+  std::map<char, std::string> const replaceMap = {
+      {' ', "%20"}, {'/', "%2F"}, {';', "%3B"}};
+
+  std::string urlEncoded = "";
+  for (auto &elem : str) {
+    if (replaceMap.find(elem) != replaceMap.end()) {
+      urlEncoded.append(replaceMap.at(elem));
+    }
+    urlEncoded.append(1, elem);
```

Comment:
there should be an else branch, otherwise the replace char isnt replaced at all

Reply:
ahh im so stupid :sweat_smile:
wonder why it was working ..